### PR TITLE
Data Skipping PR 2 of 2 - DataSkippingReader

### PR DESCRIPTION
**Pull request by scottsand-db**
This is PR 2 of 2 for Delta Lake OSS Data Skipping (aka file skipping with column stats) feature. See the issue #931.

This PR builds on the work done by #966

## What changes were proposed in this pull request?
This PR really only does two things
- finishes the implementation of DataSkippingReader
- adds DataSkippingDeltaTests test suite

### CODE REVIEW

**Comment by scottsand-db** [comment]
@tdas any other comments? else does it look good to you?

File: core/src/main/scala/org/apache/spark/sql/delta/stats/DataSkippingReader.scala [hunk 1]
```diff
@@ -51,8 +185,744 @@ trait DataSkippingReader
...
+    else {
+      // Emit filters for an imprecise range test that covers the entire entire list.
+      val min = Literal(values.min(ordering), dt)
+      val max = Literal(values.max(ordering), dt)
+      constructDataFilters(And(GreaterThanOrEqual(max, a), LessThanOrEqual(min, a)))
+    }
+  }
+
+  /**
+   * Returns a file skipping predicate expression, derived from the user query, which uses column
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/sql/delta/stats/DataSkippingReader.scala:234]: +1000... thank you for these detailed inline docs. Would have been much harder to review this large PR without this.

File: core/src/test/scala/org/apache/spark/sql/delta/DeleteSuiteBase.scala [hunk 2]
```diff
@@ -504,17 +504,14 @@ abstract class DeleteSuiteBase extends QueryTest
...
-          result)
+        checkAnswer(spark.read.format("delta").table("v"), expectResult)
       }
     }
   }
 
   testComplexTempViews("nontrivial projection")(
     text = "SELECT value as key, key as value FROM tab",
-    expectResult = Nil
+    expectResult = Seq(Row(3, 0))
```
Comment (tdas) [inline core/src/test/scala/org/apache/spark/sql/delta/DeleteSuiteBase.scala:514]: why did this change??

File: core/src/test/scala/org/apache/spark/sql/delta/test/ScanReportHelper.scala [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+import org.apache.spark.sql.delta.files.TahoeFileIndex
+import org.apache.spark.sql.delta.metering.ScanReport
+import org.apache.spark.sql.delta.stats.{DataSize, PreparedDeltaFileIndex}
+import org.apache.spark.sql.execution.{FileSourceScanExec, QueryExecution, SparkPlan}
+import org.apache.spark.sql.execution.adaptive.AdaptiveSparkPlanHelper
+import org.apache.spark.sql.execution.columnar.InMemoryTableScanExec
+import org.apache.spark.sql.test.SharedSparkSession
+import org.apache.spark.sql.util.QueryExecutionListener
+
+trait ScanReportHelper extends SharedSparkSession with AdaptiveSparkPlanHelper {
```
Comment (tdas) [inline core/src/test/scala/org/apache/spark/sql/delta/test/ScanReportHelper.scala:30]: Add docs

File: core/src/test/scala/org/apache/spark/sql/delta/test/ScanReportHelper.scala [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+   */
+  private def collectScans(plan: SparkPlan): Seq[FileSourceScanExec] = {
+    collectWithSubqueries(plan)({
+      case fs: FileSourceScanExec => Seq(fs)
+      case cached: InMemoryTableScanExec => collectScans(cached.relation.cacheBuilder.cachedPlan)
+    }).flatten
+  }
+
+  private def getListener(): QueryExecutionListener = {
+    new QueryExecutionListener {
```
Comment (tdas) [inline core/src/test/scala/org/apache/spark/sql/delta/test/ScanReportHelper.scala:45]: inline docs to explain what this listenerr does

File: core/src/test/scala/org/apache/spark/sql/delta/test/ScanReportHelper.scala [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+                partitionFilters = Nil,
+                dataFilters = Nil,
+                unusedFilters = Nil,
+                size = Map(
+                  "total" -> DataSize(
+                    bytesCompressed = Some(deltaTable.deltaLog.snapshot.sizeInBytes))
+                ),
+                metrics = scanExec.metrics.mapValues(_.value).toMap,
+                versionScanned = None,
+                annotations = Map.empty // TODO(scott): annotations
```
Comment (tdas) [inline core/src/test/scala/org/apache/spark/sql/delta/test/ScanReportHelper.scala:102]: whats the TODO for?

File: core/src/test/scala/org/apache/spark/sql/delta/test/ScanReportHelper.scala [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+                scanType = scanExec.relation.fileFormat.toString,
+                partitionFilters = scanExec.partitionFilters.map(_.sql),
+                dataFilters = scanExec.dataFilters.map(_.sql),
+                unusedFilters = Nil,
+                size = Map(
+                  "partition" -> DataSize(bytesCompressed = Some(partitionBytes))
+                ),
+                metrics = scanExec.metrics.mapValues(_.value).toMap,
+                versionScanned = None,
+                annotations = Map.empty // TODO(scott): annotations
```
Comment (tdas) [inline core/src/test/scala/org/apache/spark/sql/delta/test/ScanReportHelper.scala:121]: same qs as above

File: core/src/test/scala/org/apache/spark/sql/delta/test/ScanReportHelper.scala [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+                    bytesCompressed = Some(deltaTable.deltaLog.snapshot.sizeInBytes))
+                ),
+                metrics = scanExec.metrics.mapValues(_.value).toMap,
+                versionScanned = None,
+                annotations = Map.empty // TODO(scott): annotations
+              )
+
+              scans += report
+
+            case _ if scanExec.relation.location.rootPaths.nonEmpty =>
```
Comment (tdas) [inline core/src/test/scala/org/apache/spark/sql/delta/test/ScanReportHelper.scala:107]: what is this case for? would delta always have TahoeFileIndex? or are you planning this to be generic to non delta tables too?

File: core/src/test/scala/org/apache/spark/sql/delta/DeleteSuiteBase.scala (hunk 2 above)
Comment (scottsand-db) [reply core/src/test/scala/org/apache/spark/sql/delta/DeleteSuiteBase.scala:514]: it seems that there was a bug in Delta Lake. to be honest, I'm not entirely sure how my PR fixes that, but if you read this test case carefully it is clear that `Seq(Row(3, 0))` is the correct answer.

File: core/src/test/scala/org/apache/spark/sql/delta/test/ScanReportHelper.scala (hunk 5 above)
Comment (scottsand-db) [reply core/src/test/scala/org/apache/spark/sql/delta/test/ScanReportHelper.scala:102]: some future work if it is useful. I will remove for now

File: core/src/test/scala/org/apache/spark/sql/delta/test/ScanReportHelper.scala (hunk 7 above)
Comment (scottsand-db) [reply core/src/test/scala/org/apache/spark/sql/delta/test/ScanReportHelper.scala:107]: this case covers `parquet` tables, for example.

Comment (tdas) [reply core/src/test/scala/org/apache/spark/sql/delta/test/ScanReportHelper.scala:121]: Is this is needed for this PR? if not, i suggest you remove it. unnecessary code add clutter... we can always add this if needed for parquet in the future.

**Review by tdas — APPROVED** [review]

File: core/src/main/scala/org/apache/spark/sql/delta/stats/DataSkippingReader.scala [hunk 8]
```diff
@@ -51,8 +185,744 @@ trait DataSkippingReader
...
+  final def withStats: DataFrame = {
+    withStatsInternal
+  }
+
+
+  // Helper method for expression types that represent an IN-list of literal values.
+  //
+  //
+  // For excessively long IN-lists, we just test whether the file's min/max range overlaps the range
+  // spanned by the list's smallest and largest elements.
```
Comment (cfmcgrady) [inline core/src/main/scala/org/apache/spark/sql/delta/stats/DataSkippingReader.scala:208]: We should set a threshold, when `in.size` is less than the threshold, we should convert `In` to a series of `Equals`, and then use the filtering rules of `Equals`

For instance:


```
col_name in (v1, v2) => col_name == v1 or col_name == v2
```

> suppose `v1 < v2`

the file skipping predicate expression before:

```
stats.minValues.col_name <= v1 and stats.maxValues.col_name >= v2
```

the file skipping predicate expression after:
```
(stats.minValues.col_name <= v1 and stats.maxValues.col_name >= v1)
  or (stats.minValues.col_name <= v2 and stats.maxValues.col_name >= v2)
```

the second one has a better file filtration efficiency

File: core/src/main/scala/org/apache/spark/sql/delta/stats/DataSkippingReader.scala [hunk 9]
```diff
@@ -51,8 +185,744 @@ trait DataSkippingReader
...
+      getStatsColumnOpt(nullCountCol).flatMap { nullCount =>
+        getStatsColumnOpt(numRecordsCol).map { numRecords =>
+          DataSkippingPredicate(nullCount < numRecords, nullCountCol, numRecordsCol)
+        }
+      }
+    case Not(IsNotNull(e)) =>
+      constructDataFilters(IsNull(e))
+
+    // Match any file whose min/max range contains the requested point.
+    case EqualTo(SkippingEligibleColumn(a), SkippingEligibleLiteral(v)) =>
```
Comment (cfmcgrady) [inline core/src/main/scala/org/apache/spark/sql/delta/stats/DataSkippingReader.scala:419]: Shall we also handle the expression `EqualNullSafe`?

File: core/src/main/scala/org/apache/spark/sql/delta/stats/DataSkippingReader.scala (hunk 8 above)
Comment (scottsand-db) [reply core/src/main/scala/org/apache/spark/sql/delta/stats/DataSkippingReader.scala:208]: Hi @cfmcgrady, that's a great suggestion.

I think that this PR is big enough already ... once it is merged, would you like to implement your proposed solution? 

Or perhaps, at the very least, create an issue documenting with more detail what you'd like to implement + why it is mathematically correct (these changes can get rather complicated).

File: core/src/main/scala/org/apache/spark/sql/delta/stats/DataSkippingReader.scala (hunk 9 above)
Comment (scottsand-db) [reply core/src/main/scala/org/apache/spark/sql/delta/stats/DataSkippingReader.scala:419]: Another good suggestion. Much like the comment above, once this PR is merged, want to add this change yourself?

File: core/src/main/scala/org/apache/spark/sql/delta/stats/DataSkippingReader.scala (hunk 8 above)
Comment (scottsand-db) [reply core/src/main/scala/org/apache/spark/sql/delta/stats/DataSkippingReader.scala:208]: @cfmcgrady want to make an issue for this improvement?

File: core/src/main/scala/org/apache/spark/sql/delta/stats/DataSkippingReader.scala (hunk 9 above)
Comment (cfmcgrady) [reply core/src/main/scala/org/apache/spark/sql/delta/stats/DataSkippingReader.scala:419]: will raise a PR

**Pull request closed without merging**